inspector: Allow require in Runtime.evaluate

jkrems opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

inspector
##### Description of change

<!-- Provide a description of the change below this comment. -->

Being able to require modules from the console can be very handy while debugging issues. This allows to do it even when require is no longer in scope or the process isn't currently paused. It also creates the foundation for adding other command line API methods in the future (if more use cases come up).

jkrems commented:
I wasn't sure if this warrants a mention in the docs. So far I assume that `require` is something people will implicitly assume works (the behavior is roughly the same as in the repl / `node -e` / etc.).

Fishrock123 commented on lib/internal/bootstrap_node.js:
```diff
@@ -225,12 +225,27 @@
   }
 
+  function setupConsoleAPI(addCommandLineAPIMethod) {
+    addCommandLineAPIMethod('require', function require(request) {
```
Instead of passing, this could just access `process.inspector.addCommandLineAPIMethod` directly, whcih would be preferred.

Fishrock123 replied:
nit: I would move this below `setupGlobalConsole`

jkrems replied:
Moved `setupConsoleAPI` below and renamed it to `setupInspectorAPI` (closer to `addCommandLineAPIMethod`). Also switched to accessing the method from process instead of passing it.

jasnell commented on lib/internal/bootstrap_node.js:
```diff
@@ -249,6 +250,21 @@
   }
 
+  function setupInspectorAPI() {
+    var addCommandLineAPIMethod = process.inspector.addCommandLineAPIMethod;
```
nit: `const`?

ak239 replied:
Can we remove this method from inspector object right here to be sure that it won't leak?

jkrems replied:
Done.

jkrems replied:
Yep, will do.

jasnell commented on test/inspector/test-inspector.js:
```diff
@@ -151,6 +151,37 @@ function testInspectScope(session) {
 }
 
+function testCommandLineAPI(session) {
+  console.log('[test]', 'Verify that command line API was injected');
```
nit: I'd prefer if the test's did not include unnecessary console.log statements

jkrems replied:
Same here - should I remove the `console.log` from the other functions in this file as well? Only added it for consistency with the existing code.

jkrems replied:
Removed the `console.log` I added already.

ak239 commented on src/inspector_agent.cc:
```diff
@@ -573,6 +585,16 @@ void AgentImpl::InstallInspectorOnProcess() {
+
+void AgentImpl::AddAPIMethod(const v8::FunctionCallbackInfo<v8::Value>& info) {
+  auto env = Environment::GetCurrent(info);
+  v8::Local<v8::Object> console_api = env->inspector_console_api_object();
```
Could you check arguments here and ThrowError if something wrong in the same way like in InspectorWrapConsoleCall?

jkrems replied:
Will do.

ak239 commented on src/inspector_agent.cc:
```diff
@@ -573,6 +585,16 @@ void AgentImpl::InstallInspectorOnProcess() {
   v8::Local<v8::Object> inspector = v8::Object::New(env->isolate());
   READONLY_PROPERTY(process, "inspector", inspector);
   env->SetMethod(inspector, "wrapConsoleCall", InspectorWrapConsoleCall);
+  env->SetMethod(inspector, "addCommandLineAPIMethod", AddAPIMethod);
```
nit: I'd prefer InspectorAddCommandLineAPIMethod or AddCommandLineAPIMethod as a name of this function.

jkrems replied:
Yeah, the method name might have gotten shorter after I discovered the 80 char limit... Went back to a more descriptive name & more line breaks.

ak239 commented on src/env.h:
```diff
@@ -240,6 +240,7 @@ namespace node {
   V(domains_stack_array, v8::Array)                                           \
   V(fs_stats_constructor_function, v8::Function)                              \
   V(generic_internal_field_template, v8::ObjectTemplate)                      \
+  V(inspector_console_api_object, v8::Object)                                 \
```
Just to check, is this object accessible from JS code?

Fishrock123 replied:
this isn't an object per-se, it is not exposed to JS

jkrems replied:
As @Fishrock123 said, this is just used as a bucket of named persistent values from native code, it's never exposed to JS.

ak239 commented on lib/internal/bootstrap_node.js:
```diff
@@ -231,6 +231,7 @@
     if (process.inspector) {
       inspectorConsole = global.console;
       wrapConsoleCall = process.inspector.wrapConsoleCall;
+      setupInspectorAPI();
```
nit: I'd prefer API -> commandLineAPI. API sounds too broad for me.

Fishrock123 replied:
nah, I think this is fine.

bnoordhuis replied:
I'm with @ak239, `setupInspectorCommandLineAPI()` is a more descriptive name.

ak239 commented on src/inspector_agent.cc:
```diff
@@ -253,6 +253,7 @@ class AgentImpl {
   static void WriteCbIO(uv_async_t* async);
 
   void InstallInspectorOnProcess();
+  static void AddAPIMethod(const v8::FunctionCallbackInfo<v8::Value>& info);
```
Could you define this method outside of AgentImpl?

jkrems replied:
Moved this out and renamed (see below).

Fishrock123 commented on lib/internal/bootstrap_node.js:
```diff
@@ -249,6 +250,21 @@
 
+  function setupInspectorAPI() {
+    const addCommandLineAPIMethod = process.inspector.addCommandLineAPIMethod;
+    addCommandLineAPIMethod('require', function require(request) {
```
is this callback called synchronously? if not, I need to inspect the startup timing a bit more

jkrems replied:
The line you're commenting on is executed synchronously during bootup. From my understanding that should prevent any debug client from connecting (and calling `Runtime.evaluate`) before we registered the API method.

Fishrock123 replied:
Sounds good to me.

jkrems commented:
Sorry for force-pushing the last couple of times, missed that sentence in the contribution guide.

Pushed a commit to address the last round of feedback.

gibfahn commented:
@jkrems I don't think force-pushing is a problem, AFAIK it's a matter of personal preference. I like to use `--amend` so the commits look tidier, but either way should be fine. Which sentence in the contribution guide were you referring to?

ak239 commented:
thanks, lgtm

addaleax commented on src/inspector_agent.cc:
```diff
@@ -567,12 +578,30 @@ void AgentImpl::WaitForDisconnect() {
                            v8::ReadOnly).FromJust();                          \
   } while (0)
 
+void AddCommandLineAPIMethod(const v8::FunctionCallbackInfo<v8::Value>& info) {
```
This can be `static`, right? If so, it should probably be.

jkrems replied:
My C++ is a bit rusty, so the following might be just me missing one of the "things `static` can mean": It's no longer a class member but a top-level function as of the latest commit.

addaleax replied:
@jkrems Yeah, for top-level functions (i.e. in namespace scope, like this one), adding `static` basically means that the function will only be visible in the file in which it is defined. I think it’s generally good practice to have any top-level functions either appear in a header file to make clear that they are shared between files/part of the public C++ API, or be `static` to indicate that they are intended to be local. You can grep `src/` for `FunctionCallbackInfo<Value>& args`, most of these functions are static.

(Also, yeah, no doubt that assigning multiple distinct meanings to `static` is not an ideal thing for C++. :smile:)

bnoordhuis replied:
That's why I have a mild preference for `inline`.  It does the same thing as `static` but there is never any ambiguity as to whether it gets applied to a function or a variable declaration because it's only valid for the former.  Helpful when grepping around.

jkrems replied:
@addaleax & @bnoordhuis: Thanks a lot for the explanation. I'm pretty sure I learned & forgot this before. Once or twice.

I updated this function and also made the other two exposed C++ methods (`InspectorConsoleCall`/`InspectorWrapConsoleCall`) static so the style is consistent inside of the file. Went with `static` over `inline` because it seemed to be the dominant style in other `*.cc` files.

addaleax commented on src/inspector_agent.cc:
```diff
@@ -567,12 +578,30 @@ void AgentImpl::WaitForDisconnect() {
+  auto env = Environment::GetCurrent(info);
+
+  if (info.Length() != 2 || !info[0]->IsString() || !info[1]->IsFunction()) {
+    return env->ThrowError("inspector.addCommandLineAPIMethod takes exactly "
```
`env->ThrowTypeError`?

jkrems replied:
I thought so but it was using `ThrowError` for the same thing above. Should I change both or just my copy?

addaleax replied:
@jkrems It’s probably okay change both. :)

jkrems commented:
@gibfahn Here's [the quote I was referring to](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-6-push):

> If there are comments to address, apply your changes in a separate commit and push that to your branch.

jkrems commented:
Anything left to address here?

jkrems commented:
Rebased on `origin/master` & `make -j8 test`.

bnoordhuis commented on lib/internal/bootstrap_node.js:
```diff
@@ -257,6 +258,22 @@
+      module.filename = path.join(cwd, name);
+      module.paths = Module._nodeModulePaths(cwd);
+      return module.require(request);
+    });
```
Shouldn't `module` be defined outside the function?  I think this breaks `require('foo') === require('foo')`, i.e., that the second require returns the cached module instance.

jkrems replied:
The module cache is maintained by `Module`, so individual module instances shouldn't matter. It does influence `module.parent` (e.g. two modules that were loaded initially using the global require will not share the same `module.parent`). But that seemed less important.

Adding a test case to verify.

bnoordhuis replied:
Somebody is bound to run into that sooner or later.  If it's easy to avoid, I'd fix it now.

jkrems replied:
> Somebody is bound to run into that sooner or later.

Can you clarify what "it" is in this context? There's now a test case that verifies that it doesn't break `require('foo') === require('foo')`, unless I misunderstood your initial concern?

bnoordhuis replied:
That was my initial concern, yes, but you mentioned that the .parent property will be different for each require() call.  If that's easy to avoid, I'd do it now.

jkrems replied:
Gotcha, will add additional tests for `require('foo').parent === require('bar').parent` and fix this.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -320,6 +320,17 @@ class V8NodeInspector : public v8_inspector::V8InspectorClient {
+    for (uint32_t i = 0; i < properties->Length(); ++i) {
+      v8::Local<v8::Value> key = properties->Get(i);
+      target->Set(key, console_api->Get(key));
+    }
```
Can you use the GetOwnPropertyNames, Get and Set overloads that take a `context` as their first argument and return a Maybe/MaybeLocal?

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -483,12 +496,31 @@ void AgentImpl::WaitForDisconnect() {
+  }
+
+  v8::Local<v8::Object> console_api = env->inspector_console_api_object();
+  console_api->Set(info[0], info[1]);
```
Ditto.

bnoordhuis commented on test/inspector/test-inspector.js:
```diff
@@ -199,6 +199,36 @@ function testI18NCharacters(session) {
+      }, (message) => assert.strictEqual('object', message['result']['value'])
+    ],
+  ]);
+}
```
Can you check that:

1. `require(name)` twice returns the same object, and
2. that `delete require.cache[require.resolve(name)]` followed by require again returns a new object, and
3. that calling require again returns the same object?

`name` could be e.g. `const name = path.join(common.fixturesDir, 'empty');`

jkrems replied:
I'll try. I'm not sure if `require.cache` is exposed globally by this PR. Since the require goes through the normal `module.require`, is there any reason to believe that deleting the require cache won't affect it?

gibfahn commented:
@jkrems FYI, [this](https://github.com/nodejs/node/pull/8837#issuecomment-251140242) is no longer in CONTIBUTING.md, so feel free to squash or not, whatever you prefer...

jkrems commented:
Updated, went with adding it as another commit because consistency.

jkrems commented:
* Fixed the naming issue @bnoordhuis pointed out
* Squashed and rebased against latest master to make sure it still merges cleanly
* Reran `make test` (passes)

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -320,6 +320,19 @@ class V8NodeInspector : public v8_inspector::V8InspectorClient {
+    for (uint32_t i = 0; i < properties->Length(); ++i) {
+      v8::Local<v8::Value> key = properties->Get(context, i).ToLocalChecked();
+      target->Set(context, key,
+          console_api->Get(context, key).ToLocalChecked());
```
Can you line up the arguments and call .FromJust()?

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -483,12 +498,32 @@ void AgentImpl::WaitForDisconnect() {
+  }
+
+  v8::Local<v8::Object> console_api = env->inspector_console_api_object();
+  console_api->Set(context, info[0], info[1]);
```
.FromJust()

jkrems replied:
I hope I understood the ".FromJust" part correctly - I assume it's to make sure that a failing `Set()` actually causes an error..?

bnoordhuis replied:
Correct.

jkrems commented:
Heh, that's what I get for doing multiple things at once. They push didn't go through. *Now* the squash is pushed and the rename concern is addressed. Will have a look at the other fixes later today.

jkrems commented:
* Added test to make sure multiple globally required modules share the same parent
* Fixed style issues (aligned params / `.FromJust`)

jkrems commented:
Interesting idea. Agreed that it could be really nice to have.

darky commented:
It's alive?

jkrems commented:
Semi-alive. Kept getting side-tracked by `node inspect`. I'll see if I can make those conflicts go away, hopefully later today.

jkrems commented:
Alright, lots of churn around those parts. Nothing about the actual implementation changed but things had to move around a bit.

CI: https://ci.nodejs.org/job/node-test-pull-request/9312/

jkrems commented:
"parallel/test-zlib-failed-init" fails with some crash message on [linux-one](https://ci.nodejs.org/job/node-test-commit-linuxone/nodes=rhel72-s390x/7509/tapResults/) / [plinux](https://ci.nodejs.org/job/node-test-commit-plinux/10413/nodes=ppcle-ubuntu1404/tapResults/) / [linux-fips](https://ci.nodejs.org/job/node-test-commit-linux-fips/9880/nodes=ubuntu1404-64/tapResults/):

```tap
not ok 1444 parallel/test-zlib-failed-init
  ---
  duration_ms: 0.843
  severity: crashed
  stack: |-
    oh no!
    exit code: CRASHED (Signal: 11)
  ...
```

darky commented:
:man_facepalming:

refack commented:
> "parallel/test-zlib-failed-init" fails with some crash message on linux-one / plinux / linux-fips:

Probably related to https://github.com/nodejs/node/issues/14178

But I have 2 questions:
1. How will this behave when built `--without-inspector`?
2. Am I mixing up command line API with `node inspect`? If I'm right wouldn't this be better suited near [bootstrap_node.js#L96](https://github.com/jkrems/node/blob/6396b86c53e8ec66d71034485a7d1033962be678/lib/internal/bootstrap_node.js#L96)
anyway should be added to all processes?

jkrems commented:
> How will this behave when built `--without-inspector`?

Good point! Added an early return if `addCommandLineAPIMethod` isn't there and verified that works when using `--without-inspector`.

> Am I mixing up command line API with `node inspect`?

This has no (direct) connection to `node inspect`. The command line API is part of `v8-inspector` itself and used in all debug eval clients (e.g. the Console tab in devtools).

refack commented:
> This has no (direct) connection to node inspect. The command line API is part of v8-inspector itself and used in all debug eval clients (e.g. the Console tab in devtools).

Then why is it named `addCommandLineAPIMethod` (not gonna bikeshed over that, just a bit confusing name, if it's more like `addInspectorConsoleAPIMethod`)
But my 2nd question still stands "should it be added to all processes? anyway to add only when there is an inspector session active (something associated with the changes made by #13228)"?

jkrems commented:
> Then why is it named addCommandLineAPIMethod (not gonna bikeshed over that, just a bit confusing name, if it's more like `addInspectorConsoleAPIMethod`)

Not sure I follow - it's exposed as `inspector.addCommandLineAPIMethod`. Calling it `inspector.addInspectorConsoleAPIMethod` seems redundant..? And the setup function is called `setupInspectorCommandLineAPI`, so it also includes the inspector bit.

> But my 2nd question still stands "should it be added to all processes?"

Always adding it means it should also work when a debug session is started programmatically or after the fact (using the signal handler). Just like with the `console.*` wrappers, you don't know at the time of bootstrap if the process will ever attach a debugger.

TimothyGu commented on lib/internal/bootstrap_node.js:
```diff
@@ -307,6 +308,24 @@
+
+    addCommandLineAPIMethod('require', function require(request) {
+      return consoleAPIModule.require(request);
+    });
```
The REPL does:

```js
  var module = new Module('<repl>');
  module.paths = Module._resolveLookupPaths('<repl>', parentModule, true) || [];

  var require = internalModule.makeRequireFunction(module);
  context.module = module;
  context.require = require;
```

`_resolveLookupPaths` combines `Module._nodeModulePaths()` and `Module.globalPaths` (containing things like `~/.node_modules` and `$NODE_PATH`). We obviously don't have a `parentModule` here, but we can emulate `_resolveLookupPaths` by simply concatenating the two arrays.

`makeRequireFunction` adds things like `require.resolve`, `.main`, `.extensions`, and `.cache`, which is good for consistency as well.

Also `[consoleAPI]` isn't really useful to the user, so `<inspector console>`?

Another thing is `module`. In the REPL, `module` is available as a variable (as seen in `context.module = module` above). Such should be the case here as well if possible.

jkrems replied:
I'd like to focus on landing `require` first since I'm pretty sure that actually *wanting* to access this fake module is a fairly uncommon requirement. I'd rather see it as a potential point of confusion, especially since this "weird" module would magically get into scope as soon as there's no other `module`. I could foresee someone debugging and getting really confused that `module.exports` keeps showing an empty object even though they are *soooo* certain that they assigned properties (until they realize they didn't pause in the right spot).

As for the name - afaik it's not actually shown to the user anywhere. So I'm not sure how much value there is in bike-shedding the perfect name. :)

bnoordhuis replied:
I agree with @TimothyGu that it should at least use `makeRequireFunction()` so that `require.resolve` and friends exist.

TimothyGu commented on src/inspector_agent.cc:
```diff
@@ -711,8 +739,11 @@ void Url(const FunctionCallbackInfo<Value>& args) {
 void Agent::InitInspector(Local<Object> target, Local<Value> unused,
                           Local<Context> context, void* priv) {
   Environment* env = Environment::GetCurrent(context);
+  env->set_inspector_console_api_object(Object::New(env->isolate()));
```
Maybe set the object's prototype to `null`?

jkrems replied:
We only ever iterate over its own properties exactly once and it's not exposed anywhere else - so I don't think setting the prototype makes a difference?

TimothyGu replied:
Just seems to be the standard practice for passing objects into C++... but feel free to do whatever you prefer.

refack commented:
> And the setup function is called setupInspectorCommandLineAPI, so it also includes the inspector bit.

Was wondering what's the connection to "Command Line"


> Just like with the console.* wrappers, we don't know at the time of bootstrap if the process will ever attach a debugger.

Ack.

TimothyGu commented:
@jkrems I'm not going to block this PR based on my review above (excellent work, in fact!), but I do think consistency between the REPL and the Inspector console is important. I'll concede on `module` because it can be a point of confusion, but I can imagine people wondering "what is this `require` and why does it not have a `require.resolve`" or "why does my `$NODE_PATH` work everywhere else but not in Inspector" when using the Inspector console.

The name of the pseudo-module would be exposed through `module.id`, but since `module` is now hidden it is a non-issue.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -527,6 +527,20 @@ class NodeInspectorClient : public v8_inspector::V8InspectorClient {
   }
 
+  void installAdditionalCommandLineAPI(v8::Local<v8::Context> context,
+                                       v8::Local<v8::Object> target) {
```
Can you add the `override` keyword?

jkrems commented:
> I do think consistency between the REPL and the Inspector console is important.

That's definitely true! :) Unfortunately I'm about to be sidetracked by some work-related stuff soon so I'm not sure when I'll get around to work on those changes & the associated test cases to make sure the extended require.* API works as expected.

Given that this would put this PR into hibernation once more, I'd vote to just close this for now and cut our losses. It's a fairly minor QoL improvement and it has now consumed at least some amount of time for more than a handful of people. Maybe somebody with more experience with node who can anticipate the general feedback a bit better will be able to pick this up eventually..?

TimothyGu commented:
I'm willing to take this up and fix whatever is needed to get this in, so reopening.

TimothyGu commented:
@bnoordhuis Comments addressed, so PTAL.

CI: https://ci.nodejs.org/job/node-test-pull-request/9344/

bnoordhuis commented on lib/internal/bootstrap_node.js:
```diff
@@ -309,21 +309,19 @@
-    consoleAPIModule.paths = Module._nodeModulePaths(cwd);
+    const consoleAPIModule = new Module('<inspector console>');
+    consoleAPIModule.paths =
+      Module._nodeModulePaths(cwd).concat(Module.globalPaths);
```
Teeny tiny nit: four space indent.

TimothyGu replied:
@bnoordhuis Does the four-space indent rule apply even to JS?

bnoordhuis replied:
It's used pretty consistently in both JS and C++ code.  In fact, I thought the JS linter enforced it, even.

TimothyGu replied:
lib/buffer.js consistently uses 2-space indentation (grep for `=$`), but either way it's fine for me.

TimothyGu commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9412/

TimothyGu commented:
Rebased. I intend to land this soon if CI doesn't have any problems with that.

CI: https://ci.nodejs.org/job/node-test-commit/11573/

TimothyGu commented:
Landed in 5fd2f03b16bfd6730fe82dc3217f1d05131b9615.

gibfahn commented:
Release team were -1 on landing on v6.x, if you disagree let us know.
